clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/HelixScatterGatherRoutingStrategy.java

```diff
@@ -46,8 +47,7 @@ public List<String> getReplicas(long requestId, List<String> replicas, int requi
     return selectedReplicas;
   }
 
-  public void updateHelixGroupInfo(Map<String, Integer> helixGroupInfo) {
-    this.helixGroupInfo = helixGroupInfo;
-    this.groupIds = helixGroupInfo.values().stream().distinct().sorted().collect(Collectors.toList());
+  public void updateHelixGroupInfo(Map<String, Integer> helixGroupInfoMap) {
+    helixGroupInfoAtomicReference.set(new HelixGroupInfo(helixGroupInfoMap));
```

Comment:
I think we need to make this a `Collections.unmodifyableMap`

`updateHelixGroupInfo` is invoked in `RequestBasedMetadata` and uses the same object every time. That object is mutated with a lot of liberty. So, despite the object in this class being guarded by `AtomicReference`, the object inside gets mutated without calls to `updateHelixGroupInfo`